# Text Classification
**Instructions**:
1. Wherever you are asked to code, insert a text block below your code block and explain what you have coded as per your own understanding.
2. If the code is provided by us, execute it, and add below a text block and provide your explanation in your own words.
3. Submit both the .ipynb and pdf files to canvas.
4. **The similarity score should be less than 15%.**

##Task-1(10 points)
##What are some common applications of text classification?What are some common machine learning algorithms used for text classification?

##Answer here
**Text classification:** text classification called as a text categorization and the process of the categorize text into the group of the words using the natural language processing. text classification is defined the text and intialize the pre defined labels and categorized baesd on the previous words of the sentences or text.

**Text classification using the machine learning:**
Text classification is the machine learning techniques which initialize the tags and categoried to the documents. talking about the text classification using the natural language processing its basically analyze the text of the documents and after that its sort the text in different categories such as sentiment, topic of the documents and what is the intent behind the words is it positive or negative and also this will give the more accuracy result then humans.
**Application for text classification:**
There are lot of different application available for text classification. here.we are talking about some of them.Now, focusing on the common application in machine learning which is used for text classification :

**Spam filtering :** this is detect and filter the unrelevent to the person or unwanted  and unrequested mails,messages or comments, is they are positive or not.

**Sentiment analysis :** this will detect the data based on the emotions or sentiment in the text and identify is it positive,negative and neutral.

**Topic classification :** this will detect the topic of the document based on the data provided for instance is it related to sports, medical or crime.

**Intent classification :** this is basically to detect the purpose of the text such as whether user wants to buy product or needs the customer service.

**Language identification :** this will use for detect the language of the provided text such as english, spanish, french etc.
**Machine learning algorithms which are used for text classification:**

**1. Naive Bayes Algorithm**
This is the most popular algorithm used in the machine learning for text classification.this is the probabilistic algorithm which is use the bayes theor, and its assume that the particular class feature is present in independent of the presence of the other features.

For text classification, naive bayes is mostly used to check the probability of the text that is belongs to that particular class. its aalyze thr word frequency of the word or features in the documents and that labeled with corresponding classes.this can handle large number of features and computationally efficient, its most popular for large scaled applications.on top of that, its required less data to trained and its main goal is to get the good performace and its easily take it from the classes or domains.


**2. Support Vector Machine**
Support vector machine is the one type of the machine learning algorithm that is used for text classification. its find the hyperplane which is maximum seprates the two classes of the data points in the high dimensional spaces.In the text classification svm is use for learn decision boundry that make different pairs of the different classes of the documents based on the features.Here the features we are addressing is bag-of-words representations, TF-IDF weights, or more sophisticated representations like words embedding.Its useful with high dimensional data,where the features are represent commonly using the nlp. svm also work good with unbalanced datasets, for text classification, where the number of example in each class different significantly.Additionally, svm is able to get the highest accuracy in text classification and its mostly used in applications such as sentiment analysis, spam detection and topic based classification.

All in all, svm is the powerful machine learning techniques which can handle high dimensional data as well as the unbalanced data.


**3. Deep Learning**
Deep learning is the sunset of the machine learning which is used for text classification. basically, talking about the text classification is the process of the initialize one or  more labels to given documents based on the context of the documents. this learning model is effective in text classification because it use complex pattern and relationship in the text. this is using the multiple neurons to get the higher level features from the text.this is model is trained with large dataset so it can handle easily capture pattern and relationship.
Talking about the techniques by deep learning which is used for text clssifications are neural network as convolutional neural network(CNN), recurrent neural network(RNN), and long short-term memory(LSTM) network. this models take the text as a input and then classify the text into different categories.

All in all deep learning is the powerful for text classification, its able to achieve the goal for better accuracy on many dataset.its useful with different kind of applications which are sentiment analysis, spam detection and topic modeling.


**4. Decision Trees:** This algorithm is use the tree structure for taking the model decisions where based on the features its create the tree and nodes for getting the better accuracy this is mostly use for multi class classifications.

**5. Random Forests:**  This is work with decision trees and its basically improve the accuracy of the model and stable the predictions.


**6. Neural Networks:** this algorithm is based on the structure and function of the human brain and are useful for text classifications.









In [ ]:
# to avoid NumPy's truncation of outputs when certain code blocks are generated
import sys
import numpy
numpy.set_printoptions(threshold=sys.maxsize)

Here, we are using numpy when the truncate the output when the particulary type of code blocks generated.

here, first import the sys which used for getting functions and variables for run time environment and numpy which is here used for performing the mathematical operations.at the last numpy.set_printoptions) this is the function where numpy is set the printing options for arrays. and also passing the parameter for threshold = sys.m,axsize which is the maximum value of the variable type in the current system.and its represent the whole array print no matter the size of the array.


## **(Tutorial) Performing Naive Bayes Classification using Scikit-Learn**(5 points)

### **Bag-of-Words (BoW)**

***Bag-of-Words*** is one of the many approaches to extract features (inputs to the learning algorithm) from the text data. Depending on the basis of measure used, a Bag-of-Words representation of text contains information about the occurrence of words in the underlying text.

Of the various measures, one of them is to create a Bag-of-Words representation using the ***information about the presence/absence of words in the text***. **0** indicates the word is absent while **1** indicates the word is present.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

# 'binary' parameter set to True indicates the encoding measure is the presence/absence of words
vectorizer1 = CountVectorizer(binary=True)

# super small corpus
corpus = [
     'This is the first Document.',
     'This is the second second document.',
     'And the third one.',
     'Is this the first document?',
]

# fit the vectorizer on the corpus and then encode the data
data = vectorizer1.fit_transform(corpus)
data


<4x9 sparse matrix of type '<class 'numpy.int64'>'
	with 19 stored elements in Compressed Sparse Row format>

Here, we are vectorizing the sentences and then encoding them in to the one variable.and we are counting the vector class and convert that text into the numeric data using the  machine learning model.

Here, first we are importing the countvectorize from scikit learn. this is use for convert collection of the text into the matrix of the word counts. after that the object of the countvectorize created with parameter binary is true which means that the encoding measures is the absence of the words in the text, instead of the frewuency of the word.now we are taking the small corpus with 4 lines.after that we are using the fit_transform function with parameter as a corpus and with instace of the countvectorize and store that inot the data name variable.this method first fits he vectorizer to the corpus text and analyze them and after that it built the vocabulary of the unique words. now its convert the corpus into matrics of the words counts. and the final result is stored in the data variable.

In [ ]:
print(vectorizer1.vocabulary_.keys())    # returns the features extracted from the text (vocabulary)

data.toarray()    # returns encoded representations

dict_keys(['this', 'is', 'the', 'first', 'document', 'second', 'and', 'third', 'one'])


array([[0, 1, 1, 1, 0, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 1, 0, 1],
       [1, 0, 0, 0, 1, 0, 1, 1, 0],
       [0, 1, 1, 1, 0, 0, 1, 0, 1]])

Here, we are printing the vectorize vocabulary keys and matrix of the array counting. fisrt we are printing the vocabs from the word and after that printing the array of the matrix of this words.
 here, the vectorize1 is returns the features of the extracted from the corpus.which is the vocab of the words.this vocab is have all the unique words and tokens of the words which is available in the corpus.

now we are using the data.toarray method which returns the encoded representations of the text as the numpy array. this is row represent the text,
and the columns represent the frequency of the particular words.

overall, here we are exctracting the features from the corpus, encoding them into the numerix representation using the bag of word and returning the encoded representation as the numpy array for analysis.






In [ ]:
# 'binary' parameter when not set indicates the encoding measure is term frequency
vectorizer2 = CountVectorizer()

# again, an example corpus
corpus = [
     'This is the first Document.',
     'This is the second second document.',
     'And the third one is the document.',
     'Is this the first document document document?',
]

# same as before...
data = vectorizer2.fit_transform(corpus)
data


<4x9 sparse matrix of type '<class 'numpy.int64'>'
	with 21 stored elements in Compressed Sparse Row format>

here, we are creating the counter vector for extracting the features from the corpus. here its transform the corpus text into the matrix of the numerical values. here the binary parameter is not set, the encoding measure used in the term frequency. here, the corpus is the list of the four documents and its fit in to the fit_transform() method which is use for create sparse matrix of the text term frequency. each row is the documents and each columns is the corpus. the value in the matrix are the number of times each terms available in the each documents.

the matrix use as a input to machine learning model for tasks such as text classification and clustering.



In [ ]:


# 'binary' parameter when not set indicates the encoding measure is term frequency
vectorizer2 = CountVectorizer()

# again, an example corpus
corpus = [
     'This is the first Document.',
     'This is the second second document.',
     'And the third one is the document.',
     'Is this the first document document document?',
]

# same as before...
data = vectorizer2.fit_transform(corpus)
data


<4x9 sparse matrix of type '<class 'numpy.int64'>'
	with 21 stored elements in Compressed Sparse Row format>

here its creating the bag of words reperesentation of the corpus using the countvectorizer class from the scikit-learn. this corpus is the list of the four documents, and the countervectorize is use to convert the text into the matrix where each row is the document and column is the unique word of the corpus.

the fit_transform() method is use the countvectorize in the corpus, which first fit the corpus and then transform into the matrix of the word counts. this is stored in the data variable.

### **Using Naive Bayes in scikit-learn**

#### **Step 1. Create/Collect the Data**

In [ ]:
# importing an existing dataset from scikit-learn for demonstrating use of Naive Bayes
from sklearn.datasets import fetch_20newsgroups
dataset = fetch_20newsgroups()


# scikit-learn provides helpful utilities for out-of-the-box datasets
# what are the various categories of documents in the above dataset?
dataset.target_names

['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']

Here, we are first importing the dataset from scikit-learn existing dataset name "fetch_20newsgroups". this dataset is the collection of the news documents from 20 different kind of categories.

after fetching the dataset now we are providing the traget_name attribute which is use for retrieve the name of the categories.its return the list of the strings, where each string represent the name of the 20 different categories one by one in the dataset.

In summarize, here we are retrieving the list of the categories from the 20 news groups dataset from scikit-learn,  this is useful feature its understand the nature of the data and the classification problem which is we are tring to solve here.


In [ ]:
# what are the inputs/attributes in the above dataset?
dataset.data      # returns a list

print(f"Number of articles (inputs): {len(dataset.data)}")

Number of articles (inputs): 11314


Here, we are accesing the dataset data from the fetched 20 newsgroups. this is contain the word of the list which is present in the dataset.

here, afer that we are printing the number of articles present in the dataset by calculating the length of the data attribute. this will return the number of documents present in the dataset.

All in all, here we are accesing the data from the 20 newsgroups dataset, which is containing the text of all the documents. after that its print the number of documents in the dataset.


In [ ]:
# what are the corresponding outputs/targets in the above dataset?
dataset.target      # returns a n-dimensional NumPy array

print(f"Number of targets (outputs): {dataset.target.shape}")

Number of targets (outputs): (11314,)


Here, its accessing the target attrubute of the fetched 20 news groups dataset.target is the n-dimensional array which is containing the integer labels for all the documents available in the dataset.

here, wach integer label represent the index of the category which documents is belongs to. target_name is use for getting the list of the category names.after that it print the shape of the target array using the dataset. and its return the dimension of the numpy array, which is the number of labels available in the dataset for particular data.

In summarize, here, we are using the 20 newsgrups dataset, which is containing the interger labels represent the category for each documents belongs to. now its print the number of labels available in the dataset.


In [ ]:
# for the demonstration, let's use only a small subset from the 20 categories
categories = ['talk.religion.misc', 'soc.religion.christian', 'sci.space', 'comp.graphics']

# get the training and testing data and have them ready to use later
train_data = fetch_20newsgroups(subset='train', categories=categories)
test_data = fetch_20newsgroups(subset='test', categories=categories)
print(train_data.data[5])

From: dmcgee@uluhe.soest.hawaii.edu (Don McGee)
Subject: Federal Hearing
Originator: dmcgee@uluhe
Organization: School of Ocean and Earth Science and Technology
Distribution: usa
Lines: 10


Fact or rumor....?  Madalyn Murray O'Hare an atheist who eliminated the
use of the bible reading and prayer in public schools 15 years ago is now
going to appear before the FCC with a petition to stop the reading of the
Gospel on the airways of America.  And she is also campaigning to remove
Christmas programs, songs, etc from the public schools.  If it is true
then mail to Federal Communications Commission 1919 H Street Washington DC
20054 expressing your opposition to her request.  Reference Petition number

2493.



here, we are using the small dataset of 20 newsgropus data. and here we are selecting the categories such as 'talk.religion.misc', 'soc.religion.christian', 'sci.space', and 'comp.graphics'.

now, fetch 20 news groups function is called again with the parameter set to train to fetch the training data and also with the parameter set as test to fetch the testing data.here the category parameter is set to filter the dataset and only include the documents which is belongs to that category.

After gettung the training data, here its use the index operator to access the text of the sixth document using the train data.data[5]. the text of this documents is printed.at

In conculsion, we are selecting the small amount of the data from the 20 newsgroups dataset and fetch the training data for this categories. at the last we are printing text of the sixth document in the training data.



#### **Step 2(a): Prepare the inputs for modeling**

In [ ]:
# create BoW representations for the training data (excluding targets); aka feature extraction
from sklearn.feature_extraction.text import CountVectorizer
bow_vectorizer = CountVectorizer()

# first, build the training vocabulary, fit()
# then, use the vocabulary to transform the training data into a document-term matrix, transform()
# results in a document-term matrix structure
X_train = bow_vectorizer.fit_transform(train_data.data)

# let's check the document-term matrix
X_train

# convert sparse matrix to dense matrix
# X_train.to_array()

<2153x35329 sparse matrix of type '<class 'numpy.int64'>'
	with 352179 stored elements in Compressed Sparse Row format>

Here, first we are importing the countvectorize for extracting the features form the documents.here we are using the x_train.to_array to convert the sparse matrix X_train into a dense matrix. sparse matrix is where the  is having elements are zero while dense matrix is where almost all of the elements are non zero. here we are using the toarray method which is use for convert the sparse matrix to a dense matrix  create array with the same shape uisng the sparse matrix its fill the non-zero elements.

All in all, the X_train.to_array() is convert the sparse matrix intot he normal matrix and its display easily and manipulated.


#### **Step 2(b): Prepare the outputs for modeling**

In [ ]:
# store the outputs (corresponding to the news articles in the training set) for easy access
y_train = train_data.target     # returns a n-dimensional NumPy array
y_train

# print(f"y_train array size: {y_train.shape}")

array([2, 2, 1, 2, 0, 3, 1, 0, 3, 3, 2, 0, 1, 2, 3, 2, 2, 0, 1, 0, 1, 0,
       1, 2, 2, 0, 0, 1, 3, 3, 1, 1, 2, 1, 1, 1, 0, 2, 0, 3, 2, 1, 2, 2,
       2, 1, 1, 1, 1, 2, 3, 3, 1, 0, 1, 2, 3, 2, 2, 2, 0, 3, 1, 1, 2, 3,
       0, 2, 0, 0, 1, 0, 2, 1, 2, 1, 0, 0, 1, 1, 2, 2, 1, 0, 2, 2, 0, 2,
       0, 2, 0, 2, 1, 1, 3, 0, 1, 1, 0, 1, 0, 3, 0, 2, 3, 0, 2, 2, 3, 3,
       1, 3, 2, 1, 0, 2, 0, 1, 2, 0, 2, 3, 0, 1, 2, 2, 1, 2, 2, 1, 3, 2,
       1, 0, 2, 1, 0, 0, 2, 1, 2, 1, 0, 2, 1, 1, 2, 0, 1, 2, 2, 1, 0, 2,
       2, 3, 2, 0, 1, 0, 0, 1, 3, 2, 2, 1, 2, 2, 0, 3, 2, 0, 2, 0, 0, 0,
       2, 0, 0, 2, 0, 2, 3, 2, 2, 0, 3, 1, 1, 1, 1, 1, 1, 3, 1, 0, 2, 3,
       2, 3, 3, 2, 1, 0, 1, 3, 1, 3, 2, 0, 2, 1, 1, 3, 0, 1, 1, 2, 1, 2,
       2, 3, 0, 1, 1, 1, 3, 0, 3, 0, 2, 1, 0, 0, 0, 1, 0, 2, 2, 1, 2, 1,
       2, 0, 3, 1, 1, 1, 1, 1, 2, 2, 3, 3, 2, 1, 0, 2, 0, 0, 2, 3, 0, 2,
       0, 2, 1, 3, 3, 3, 1, 0, 2, 1, 3, 0, 2, 2, 3, 2, 0, 0, 2, 2, 1, 1,
       0, 0, 2, 2, 0, 1, 1, 2, 0, 0, 0, 1, 3, 2, 1,

here, we are storing the largest values of the news in the training set in to the numpy array y_train. here target value is present the categories or labels for the news.

talking about the machine learning, we are spliting the data intot he training and testing sets. where we use the training set to train the model while testing data will use to evaluate or testing the model and check how model is performing with unseen data.

here, y_train has a shape n where n is the numeber of news articles in the training data. the shape represent that here the array is one dimensional. here we print the array, which is the number of the elements in the array.


##### **Transforming the outputs for a supervised learning task**

**PLEASE READ:** Depending on the dataset you are working with, there could be datasets that contain non-numerical labels corresponding to the targets (outputs for a supervised learning problem). scikit-learn also includes utilities that you could use to convert (transform) your non-numeric labels to numeric ones.

In [ ]:
from sklearn import preprocessing
tgt_enc = preprocessing.LabelEncoder()

# assume the following is the list of unique classes in your data
some_data_targets = ["paris", "paris", "tokyo", "amsterdam", "paris", "tokyo", "tokyo", "tokyo", "amsterdam", "england"]

# fit your targets of the training data to the LabelEncoder instance
tgt_enc.fit(some_data_targets)

# get the set of unique classes
print(f"Unique categories: {list(tgt_enc.classes_)}")

# encode the targets as numerical labels
encoded_tgts = tgt_enc.transform(["tokyo", "tokyo", "paris"])
print(f"Encoded labels: {encoded_tgts}")

Unique categories: ['amsterdam', 'england', 'paris', 'tokyo']
Encoded labels: [3 3 2]


Here, we are using the labelencoder class from the sklearn. here we are preprocessing module in python to encode categorical target values into the numeric labels.Here we are importing the labelencoder class and instance of the class or say object of the class created.

after that we are defining the list of the target data name as some_data_targets. here we are fiting the targets data from the training set into the labelencoder objects and for fit the data fit() method is used.

after that class attribute tgt_enc is use to get the unique classes. this procedure will be done by classes_attribute and convert this into the list.

at the last we are using the transform method which encode the new categorical data into the numerical labels.here three cities tokyo,tokyo,paris will be encoded into the numeric form using the transform() method. at the last we are sorting the results and print them using the encoded_tgts.


#### **Step 3: Building a Learning Model Using Naive Bayes Algorithm**

** **IMPORTANT!** ** - Regardless of the task, when building a learning model, always make sure ONLY the data from training set is used to train the model. ***Testing set MUST NEVER to be used to train/build the model***. Testing set is used only to report the results of your model, which is the last step of the process (after the model is trained and you have found a best model).

**Note:** Building the model is also referred to as training the model.

In [ ]:
from sklearn import preprocessing
tgt_enc = preprocessing.LabelEncoder()

# assume the following is the list of unique classes in your data
some_data_targets = ["paris", "paris", "tokyo", "amsterdam", "paris", "tokyo", "tokyo", "tokyo", "amsterdam", "england"]

# fit your targets of the training data to the LabelEncoder instance
tgt_enc.fit(some_data_targets)

# get the set of unique classes
print(f"Unique categories: {list(tgt_enc.classes_)}")

# encode the targets as numerical labels
encoded_tgts = tgt_enc.transform(["tokyo", "tokyo", "paris"])
print(f"Encoded labels: {encoded_tgts}")

Unique categories: ['amsterdam', 'england', 'paris', 'tokyo']
Encoded labels: [3 3 2]


Here, we are using the labelencoder class from the sklearn. and its use for convert the categorical data into the numerical labels. the labelencoder is use for encoding the categorical features as an integer array.

first, we create the instance of the labelencoder class, then we are fiting the encoded training data by calling fit() method on the object of the encoder using the some_data_target_list. here trains the encoder to identify the unique classes of the data.
after that class_attribute we are using to retrieve the unique classes which is recognize during the training time of the model. its return the sorted list of the unique classes.

At the last, we are using the method transform() which is encode object use to convert categorical data into the numberical data. here encoder convert the"tokyo","tokyo","paris" into the numerical lebels. here the output is store in the variable encoded_tgts and print them.

overall, here we are using the labelencoder to process the categorical data for machine learning models.





#### **Step 3: Building a Learning Model Using Naive Bayes Algorithm**

** **IMPORTANT!** ** - Regardless of the task, when building a learning model, always make sure ONLY the data from training set is used to train the model. ***Testing set MUST NEVER to be used to train/build the model***. Testing set is used only to report the results of your model, which is the last step of the process (after the model is trained and you have found a best model).

**Note:** Building the model is also referred to as training the model.

In [ ]:
from sklearn.naive_bayes import MultinomialNB
mnb_model = MultinomialNB()
mnb_model.fit(X_train, y_train)

MultinomialNB()

Here. we are first importing the sklearn library for implementing the naive bayes algorithm, specially, multinomial naive bayes  to train the machine learning model on  the labeled data.

after that we are creating the instance of the multinomialNB , which is use to fit the model in the training data.here we are using the fit() method to used to train the multinomialNB model using the training data. X_train present the traing data of the model, while Y_train data is the test data which is used for evaluate or train the model.

all in all. here we are fiting the multinomialNB naive bayes model train using the training data, once the model is trained, its use for prediction for the unseen data as well as the new data.




#### **Step 4(a): Make Predictions Using the Trained Model**
**Note:** We are assuming that the above model trained is the best model for the data under consideration. *In reality, decision about the best model is based on performing hyperparameter tuning on a tuning/validation data set*. While hyperparameter tuning is out of scope for this notebook, you can always lookup articles, blog posts about this topic on the World Wide Web.

In [ ]:
# before testing the model, ensure your test data, both inputs and outputs, is ready for use:

# 1. preprocess your testing data into a document-term matrix (using the training vocabulary)
X_test = bow_vectorizer.transform(test_data.data)
X_test

<1432x35329 sparse matrix of type '<class 'numpy.int64'>'
	with 230051 stored elements in Compressed Sparse Row format>

Here, we are preparing the testing data for the machine learning model and its trained on the bag-of-words present the training data.

here. we are uses the bow vectorizer object which was previously created and fit the traing data which transform the test data into the text matrix. this is represent the frequency of the occurance of the each word in the test documents,here its use the same vocab as the training data.

transformation is use for test data in the same format as the training data and its use as the input of the machine learning model it was trained using the bow representations.


In [ ]:
# 2. store the known outputs corresponding to the articles in the test set
y_test = test_data.target

print(f"y_test array size: {y_test.shape}")

y_test array size: (1432,)


Here. we are using the test data for the dataset and split the test set for evaluating the model.

and y_test= test_data.target is use for assign the variable. for instance, the ouput or label. here y_test is one dimensional array which contains the output of the corresponding documents in the test data.

after that we are printing the y_test array size using the shape which is the number of the elements of the array. here its shows that how many articles in the test data where the how many known outputs are available for evaluating the perfomance of the model.


In [ ]:
# finally, apply the model on the test set to make predictions
# in this case, predictions are classification labels
predictions = mnb_model.predict(X_test)

here, we are predicting the training model to the test set and here we are making the prediction of the target variables.for instace, we are classifying the articles in to the different categories.

here we are using the mnb_model for trained the naive bayer model and we are fitting the model using the training data and after that we are using the predict() method for the naive bayes model that takes input as the X_train and return the prediction values which is the predicted classification label for each documents.

we are using the prediction , which is one dimensional array which is containg the predicted label of the each documents for test set. predicted labels will be compared to the known outputs to y_tests for evaluating the performance of the model for test set.



In [ ]:
# using the same model but predictions are probabilities
y_pred_prob = mnb_model.predict_proba(X_test)
print(y_pred_prob)

[[2.75544189e-101 1.35696571e-093 9.04229140e-034 1.00000000e+000]
 [1.00000000e+000 8.11913323e-031 5.82342828e-045 1.00498728e-040]
 [7.23391406e-081 1.00000000e+000 6.57667626e-091 2.33963648e-074]
 [1.00000000e+000 2.06238640e-062 6.13232665e-085 3.97450971e-090]
 [4.78607486e-029 9.99999460e-001 5.40232163e-007 7.64770874e-016]
 [1.97449728e-016 3.80488166e-013 1.00000000e+000 1.31095855e-013]
 [1.00746827e-222 1.00000000e+000 1.75010680e-256 2.36523385e-255]
 [1.24078514e-069 1.00000000e+000 1.14014415e-084 1.21604211e-078]
 [1.00000000e+000 3.13263274e-015 1.72011457e-017 6.44230160e-014]
 [3.39809835e-025 1.00000000e+000 2.57276712e-039 1.83202609e-037]
 [9.21847391e-033 3.74504638e-027 2.58294255e-015 1.00000000e+000]
 [4.44624717e-012 8.12823934e-012 9.99999268e-001 7.32111989e-007]
 [1.15879118e-054 1.88240173e-035 1.20272394e-031 1.00000000e+000]
 [2.98126336e-165 2.98083778e-158 5.95385832e-006 9.99994046e-001]
 [6.53315939e-051 1.25933135e-047 9.99999955e-001 4.51282503e-

here. first we are testing the model using the test data y_test to make the prediction of the target variable.

here we are using the pridict_prob() method with naive bayes model  that take input as the X_test and return the array of predicted probability for each possible classification label for each documents.

here we are assigning the output of the predict_prob() intot he y_pred_prob which is the two dimensional array which is containing the predicted probabilities of the each articles in the test set for each documents set the classification labels.

after that we are printing the predicted values and shows the probaility of the prediction values for each articles in the test set. we are evaluating the model using the test data and checking how well the model is in predicting the labels.



#### **Step 4(b): Evaluate the Model**

In [ ]:
# importing the metrics module from sklearn
from sklearn import metrics

# use the accuracy_score metric to calculate accuracy of the model
# you evaluate a model by comparing its predictions against the known outputs of the test set
accuracy = metrics.accuracy_score(y_test, predictions)

Here, we are importing the sklearn  matrix which contains the different kind of functions for evaluating the performance of the machine learning model.

Here we using the accuracy= metrics.accuracy_score(Y_test, predictions) for checking the accuracy of the model and check the performance of the model that how well its predicting using the testing data or we can say unseen data.


here accuracy score take known output y_test and predicted output of the predictions in input and returns the accuracy of the model.





In [ ]:
# use the confusion matrix metric to understand the predictive power of the model
print(metrics.confusion_matrix(y_test, predictions))

[[371  11   2   5]
 [ 11 377   5   1]
 [  5   4 379  10]
 [  5  11  49 186]]


here we are printing the confusion matrix using the y_test and using the model fit object prediction. confusion matrix is the table for summarizing the performance of the classification model  in the test data.

here the confusion matrics provides the more understanding the performance of the model than the accuracy score. its use for calculate the different evaluation matrix like precision,recall and F1-score.this all are useful for checking the performace of the model.



##Task-2(20 points)
##Create a dataset of email samples and their corresponding labels. Convert the email samples into TF-IDF vectors and train a Naive Bayes classifier using this dataset. Compare the performance of this model with the model trained in the previous question.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Load the dataset from Kaggle
spam_df = pd.read_csv('/content/Email_spam.csv')

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(spam_df['text'], spam_df['spam'], test_size=0.2, random_state=42)

# Convert the email samples into TF-IDF vectors
tfidf_vectorizer = TfidfVectorizer(stop_words='english')
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
X_test_tfidf = tfidf_vectorizer.transform(X_test)

# Train a Naive Bayes classifier using the dataset
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train_tfidf, y_train)

# Evaluate the performance of the model on the testing set
y_pred = nb_classifier.predict(X_test_tfidf)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred, pos_label=1))
print("Recall:", recall_score(y_test, y_pred, pos_label=1))
print("F1 score:", f1_score(y_test, y_pred, pos_label=1))


Accuracy: 0.8926701570680629
Precision: 1.0
Recall: 0.5758620689655173
F1 score: 0.7308533916849016


**Naive Bayes Classification :** naive bayes classifier is the algorithm for probability classification where this classification is predict the model use the probablistic way.

**TF-IDF :** TF-IDF stands for term frequency - inverse dense frequency this is technique use for searching the meaning of the sentences and its cancel the bag of words which are not capable for the text classification for model read word in numbers.


here, we are using the email dataset for classify the emails that they are spam or not spam. here we are loading the dataset using the pandas library for read the csv file.and we are here splits the training and testing the dataset and its convert the emails into the TF-IDF vectors using the TFidvectorizer from the sklearn.

here we are training the model using the naive bayes classifiers using the multinomialNB from the sklearn, and evaluates the performance of the classifier using the test data using the different evaluation matrix from the sklearn.here for evaluation we are using the accuracy score, precision score, recall score and f1 score.

here we are calculating the accuracy score of the model on the test data, which is proportion of correct prediction by the model.here it takes input as the Y_test and y_pred value as well as the position label =1.

after that we are calculating the precison score using the parameter y_test, y_pred and pos_label =1 for test the model, which is proportion of the true positive prediction by model.

Additionally, we are calculating the recall_score too using the parameters y_test,Y_pred and pos_label=1 for test the model, which is the proportion of the true positives from all the actual positive examples in the test data.

after that we are calculating the f1 score using the parameter y_test, y_pred and pos_label=1, which is the harmonic mean of the precision and recall.

here the evaluation matrics is provide the performance of the naive bayes classification using the email spam classifiers. here we are comparing the metrics, where we can identify the model performance on the email whether email is spam or not.here the accuracy of the model is 0.89.




#Task 3 (15 points)

##How does the Naive Bayes classifier work in the context of text classification?What is the difference between bag-of-words and term frequency-inverse document frequency (TF-IDF) approaches in text classification?

##Answer here
**Naive Bayes Classification:** this is the probabilistic model which is use the bayes theorm for classify the text depends on the probabilities of each word which is available in the each class.

**Context of text classification:**
Text classification is the process of categorise the text into different classes.Talking about that how the naive bayes work, let's assume that we have text and we want to classify them into two different classes name as "spam" and "Not Spam". here the naive bayes classification will calculate the probability of the each word. for instance that "Deal" word is comes in the spam class as 0.6 while in "Not spam" it comes only 0.1. after that here we are inputing the new text, now the  model will calculate the probability of the text to each class by multiplying the probability of each word in the text which are available in each class. at the last model assign the class with the high probability.

**Bag-of-Words:**
bag-of-word is use for presenting the text depends on the frequency of the words that appear in the documents. this method is creating the list of the all the words that will appear in the documents, and at the last its count that how many number of times that word is appear in the text.

**TF-IDF:**
TF-IDF stands for term frequency-inverse document frequency. This is the technique for information retrieval and machine learning to determine the significance of the word, phrases and string present in the document its compare to the other documents. its gives us the numeric value which represent that how many time this word appear in the document as well as count the unique words too. which is useful for identifying the useful and relevant term in the collections.

**Difference between between bag-of-words and TF-IDF approach in text classification**
here, we are going to discuss about the difference between bag-of-word and TF-IDF approaches in the text classification.

Talking about the bag-of-words approach it represent the each text in to the words, and also its ignore the order of word as well as the context of the words. for instance, if the sentence is "I like to eat pizza, pizzas are my favorites" so here in the bag-of-words its represent such as {I,like,to,eat,pizza,are,my,favorites}. then each text is present as the vector where each of the element is represent the frequency of the each words.

In contrast, the TF-IDF approach its not only take the frequency of the each word but also that word is common in whole text. the approach is the word its appear in the text frequently but rarely its use in the other documents which is more important to differentiate between classes.

talking about how TF-IDF calculated, here TF(term frequency) which is represent the frequency of the word. its calculated based on the number of times the word appear in the text divided by the total number of the words.And Inverse document frequency is calculate based on the common word across the all document in the dataset. its basically calculate using the logarithm of the total number of texts in the dataset divided by the number of text that have that words. here TF-IDF is the product of the TF and IDF.

while in TF-IDF approach, each text is present as a vector which is each element that represent the TF-IDF score for each of the words. this is mostly used in the text classification because its give the better performance compared to bag-of-words,when working with the large and diverse dataset.





#Note:
For Tasks 4 and 5, you can use any of the in-built datasets like iris dataset, boston dataset, diabetes dataset etc. wherever you are asked to use your own dataset.

##Logistic Regression

Logistic regression is an example of a discriminative classifier and is commonly used in text classification, as a baseline in research, and as an MVP in real-world industry scenarios.


##Task-4(20 points)
###Tutorial: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

###Implement the Logistic Regression classifier in the link given above on a dataset of your choice. You can implement the example section in the tutorial link given above.

In [ ]:
from sklearn.datasets import load_breast_cancer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# load the breast cancer dataset
cancer = load_breast_cancer()

# split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, test_size=0.3)

# create a Logistic Regression classifier
clf = LogisticRegression()

# train the classifier on the training set
clf.fit(X_train, y_train)

# make predictions on the testing set
y_pred = clf.predict(X_test)

# calculate the accuracy of the classifier
accuracy = clf.score(X_test, y_test)
print(f"Accuracy: {accuracy}")

Accuracy: 0.9532163742690059


/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


**Logistic Regression Classification:**
This is the algorithm for natural language processing which is use for the classification, which means its use for the binary classification when the data is binary then this algorithm is use for classify. and the output is define as the 0 and 1.

Here, we are using the logistic regression classification for solving the issue of binary data. here we are using the breast cancer dataset from the sklearn library which is use for predict the breast tumor is malignant or benign.

here, we first split the dataset intot he training and testing using the train_test_split() function. here the training data us used to train the model while the testing data is used for evaluating the model or checking the performance of the model.

Here, we are creating the classifier using the logisiticregression() function. here we are using the fit() method to train our model using the training data.

After that we are using the predict() method to check the performance of the model. we are using the test set for evaluate the model.and also we are comparing the predictions with the true labels using the scale() method, which is return the calculate accuracy of the classifier.At the last we are now printing the accuracy score with breast cancer dataset is 0.97.

All in all, here we are using the logistic regression for binary classify the breast cancer dataset.





In [ ]:
from sklearn.datasets import fetch_openml
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# load the MNIST dataset
mnist = fetch_openml(name='mnist_784')

# split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(mnist.data, mnist.target, test_size=0.3, random_state=42)

# create a Logistic Regression classifier
clf = LogisticRegression()

# train the classifier on the training set
clf.fit(X_train, y_train)

# make predictions on the testing set
y_pred = clf.predict(X_test)

# calculate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")


/usr/local/lib/python3.9/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(
/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Accuracy: 0.9196666666666666



here is the instance of the multi-class classification using the logistic regression.here we are using the MNIST dataset from the sklearn library for recognize the handwritten digits.

Here, first we are loading the dataset using the fetch_openml() function.after that we are spliting the data into training and testing using the train_test_split() function.here training data is used to train the model while the testing data is used to test the model or evaluate the performance of the model.

Alogistic regression classifier is here created using the logisticregression() function. also here we are using the fit() method to train the model using the training data.

after that we are using the predict() method for that we are using the testing data. now this prediction compared to the true labels using the accuracy_score()function, which is calculates the accuracy of the model.

at the last here we are printing the accuracy using the logistic regression model for MNIST dataset which is 0.91.

All in all, here we are training and testing the multi-class classification model using the logistic regression classifier for MNIST dataset.






In [ ]:
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
X, y = load_iris(return_X_y=True)
clf = LogisticRegression(random_state=0).fit(X, y)
clf.predict(X[:2, :])

clf.predict_proba(X[:2, :])


accuracy=clf.score(X, y)
print(f"Accuracy: {accuracy}")


Accuracy: 0.9733333333333334


/usr/local/lib/python3.9/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Here, we are doing the binary and multi-class classification issues using the logistic regression model. here we are using the Iris dataset from the sklearn library to classify the Iris flower using its features.

here first the Iris dataset is loaded using the load_iris function. this return the  input features as x and target labels as a y. the x variable have some measurement of the petals and sepals of the flowers, in contrast, y is having the corresponding species labels.

here the classifier is created using the logisticregression() function.here we are using the fit() method to train the model on the entire dataset.and after that we are using the predict() method to make the prediction on the first two samples of the dataset.

Here, we are using the predict_prob() method for calculating the probability of the each class for the first two samples.here this method is use for assess the confidence of the model for its predictions.

here,we are calculating the accuracy_Score() method which is use to calculate the classifier performance for the whole dataset. here we using the score() method for classifier to calculate the accuracy. here its present the percentage of the correctly classify the instances of the dataset.

all in all, here we are training and testing the logistic regression classifier on the MNIST dataset. the main aim is to accurately classify the Iris flowers with its features using the supervised learning approach. the accuracy score for that is 0.97.



#Task 5 (20 points)
###Tutorial:https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html

###Implement Decision tree classifier on a dataset of your choice. Make use of the tutorial link given above.

In [ ]:
##your code here
from sklearn.datasets import load_breast_cancer
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# load the breast cancer dataset
cancer = load_breast_cancer()

# split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, test_size=0.3, random_state=42)

# create a Decision Tree classifier
clf = DecisionTreeClassifier()

# train the classifier on the training set
clf.fit(X_train, y_train)

# make predictions on the testing set
y_pred = clf.predict(X_test)

# calculate the accuracy of the classifier
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")


Accuracy: 0.9239766081871345


**Decision classifier:**
Decision tree classification is the process of solving the classification and regression problems using the rules. here its split the data based on the values of the feature until the all data point will be group together. this is same as the flowchart , where each node present as a feature based split. that whole process starts with the root node and ends with the decision made at the leaves.this is use for both classification as well as the regression.

Here, we are using the sklearn library to create the decision tree classifier for the breast cencer dataset.

Here, we are first loading the dataset of the breast cancer using the load_breast_cancer() function from the sklearn library.here we are spliting the data into the training and testing dataset using the train_test_Split() function.
here, now the we are creating decision tree classifier using the decisiontreeclassifier() function from the sklearn. here its train using the method fit(). after that we are using the predict() method to make the prediction using the testing data.
 at the last we are calculating the accuracy of the clssifier using the accuracy_Score() function. and here we are printing the accuracy score which is 0.92.

The  main aim is to do that we are identifying that how to use the sklear library for make and evaluate the decision tree classification for the breast cancer dataset. this is one type of the supervised learning algorithm use for decision tree as well as the regression. here for training the decision tree classifier we are using the breast cancer data for predict whether the tumor is malignat or benign based on the features of the tumor.  the accuracy is calculated and classify the model whether its predict well or not.



In [ ]:
from sklearn.datasets import load_iris
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
clf = DecisionTreeClassifier(random_state=0)
iris = load_iris()
cross_val_score(clf, iris.data, iris.target, cv=10)

array([1.        , 0.93333333, 1.        , 0.93333333, 0.93333333,
       0.86666667, 0.93333333, 1.        , 1.        , 1.        ])

here , we are doing the cross validation using the iris dataset for the decision tree classifier, after that here we load the dataset using the load_iris() function from the sklearn. this classifier is set as the random state which is 0.

here after that we are using the function cross_val_score() form sklearn which is used to perform the k-fold cross_validation, where we are setting the k=10. this is return the score of the array, where each score is the accuracy of the classifier on the different part of the dataset.

the main aim is to cross validation is to evaluate the model that how well the classifier is performing in the new data or unseen data. here we split the data into the multiple part and train the classifier on different combination of the dataset parts. after that we are getting more accurate estimationof the classifier performance than the performance of the simple of classify model with the train and testing data.


# **Text classification using the three different algorithms naive bayes, decision trees and logistic regression**

In [ ]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Load the 20 Newsgroups dataset
newsgroups_train = fetch_20newsgroups(subset='train')
newsgroups_test = fetch_20newsgroups(subset='test')

# Convert the text data to TF-IDF vectors
tfidf = TfidfVectorizer()
X_train = tfidf.fit_transform(newsgroups_train.data)
X_test = tfidf.transform(newsgroups_test.data)

# Train and test Naive Bayes classifier
nb = MultinomialNB()
nb.fit(X_train, newsgroups_train.target)
nb_pred = nb.predict(X_test)
nb_acc = accuracy_score(newsgroups_test.target, nb_pred)
nb_prec = precision_score(newsgroups_test.target, nb_pred, average='weighted')
nb_rec = recall_score(newsgroups_test.target, nb_pred, average='weighted')
nb_f1 = f1_score(newsgroups_test.target, nb_pred, average='weighted')
print("Naive Bayes:")
print(f"\tAccuracy: {nb_acc}")
print(f"\tPrecision: {nb_prec}")
print(f"\tRecall: {nb_rec}")
print(f"\tF1-score: {nb_f1}")

# Train and test Logistic Regression classifier
lr = LogisticRegression()
lr.fit(X_train, newsgroups_train.target)
lr_pred = lr.predict(X_test)
lr_acc = accuracy_score(newsgroups_test.target, lr_pred)
lr_prec = precision_score(newsgroups_test.target, lr_pred, average='weighted')
lr_rec = recall_score(newsgroups_test.target, lr_pred, average='weighted')
lr_f1 = f1_score(newsgroups_test.target, lr_pred, average='weighted')
print("Logistic Regression:")
print(f"\tAccuracy: {lr_acc}")
print(f"\tPrecision: {lr_prec}")
print(f"\tRecall: {lr_rec}")
print(f"\tF1-score: {lr_f1}")

# Train and test Decision Tree classifier
dt = DecisionTreeClassifier()
dt.fit(X_train, newsgroups_train.target)
dt_pred = dt.predict(X_test)
dt_acc = accuracy_score(newsgroups_test.target, dt_pred)
dt_prec = precision_score(newsgroups_test.target, dt_pred, average='weighted')
dt_rec = recall_score(newsgroups_test.target, dt_pred, average='weighted')
dt_f1 = f1_score(newsgroups_test.target, dt_pred, average='weighted')
print("Decision Tree:")
print(f"\tAccuracy: {dt_acc}")
print(f"\tPrecision: {dt_prec}")
print(f"\tRecall: {dt_rec}")
print(f"\tF1-score: {dt_f1}")


Naive Bayes:
	Accuracy: 0.7738980350504514
	Precision: 0.8218781741893993
	Recall: 0.7738980350504514
	F1-score: 0.7684457156894653
Logistic Regression:
	Accuracy: 0.8274030801911842
	Precision: 0.8306575949769146
	Recall: 0.8274030801911842
	F1-score: 0.8257218194818299
Decision Tree:
	Accuracy: 0.555231014338821
	Precision: 0.5591830480724167
	Recall: 0.555231014338821
	F1-score: 0.5556692497281566


here we are performing the text classification using the 20 newsgroups dataset using the three different kind of classifier which is Naive bayes, Logistic regression and Decision tree.

here we are using the 20 newsgroups dataset which contain the 20000 newsgroups posts, and we are parting them into equal part with 20 different newsgroups. here we are loading the dataset using the fetch_20newsgroups() function from the sklearn. we now split the dataset into the training and testing the data and convert the text data to TF-IDF vectors using the Tfidfvectorizer() function from the sklearn.

here, we are training the three different classifier using the training data, naive bayes,logistic classification and decision tree. we are using the multinomialNB(), logisticregression() and decisiontreeclassifier() from sklearn to create the classifiers and use the fit() method to train the model.

here after that we are predicting the model using the predict() method using the testing data for each classifier, and calculating the different evaluate the matrics such as accuracy, precision, recall and f1 score function from the sklearn. here along with that we are using the average=weighted  parameter to calculate the weighted average of the metrics across all the classes.

at the last we are printing the evaluation matrics of the each classifier such as naive bayes, logistic regression, decision tree and the accuracy score are 0.77,0.82,0.55 respectively.


##Task -6(10 points)
##Compare your results of Naive Bayes, Logistic regression and Decision Tree. Which one do you think is the best text classifier?

##Answer here
**Naive bayes:**
this is the probabilistic model which is use the bayes theorm for classify the text depends on the probabilities of each word which is available in the each class.naive bayes is the probabilistic algorithm which is use the conditional probability.
**Logistic regression:**
This is the algorithm for natural language processing which is use for the classification, which means its use for the binary classification when the data is binary then this algorithm is use for classify. and the output is define as the 0 and 1. logistic regression model is the linear model where the event occured based on the independent variables.
**Decision tree:**
Decision tree classification is the process of solving the classification and regression problems using the rules. here its split the data based on the values of the feature until the all data point will be group together. this is same as the flowchart , where each node present as a feature based split. that whole process starts with the root node and ends with the decision made at the leaves.this is use for both classification as well as the regression.


**Comparing the result of the different classifiers**
the effectiveness of the classifier based on the specific dataset. however that logistic regression and naive bayes is used for the text classification while decision tree commonly use for classification for structured data.

here in the example above we are using the 20newsgroups dataset to classify and the accuracy score, precision, recall and f1 score as follows
**Naive Bayes:**
	Accuracy: 0.77
	Precision: 0.82
	Recall: 0.77
	F1-score: 0.76
**Logistic Regression:**
	Accuracy: 0.82
	Precision: 0.83
	Recall: 0.82
	F1-score: 0.82
**Decision Tree:**
	Accuracy: 0.55
	Precision: 0.55
	Recall: 0.55
	F1-score: 0.55

Focusing on this results for the performance of the classifiers and its shows that naive bayes and logistic regression work well with the 20newsgroups dataset for the text classification. while, decision tree model perform poor compared to the naive bayes and logistic regression with a lower accuracy_score, precision,recall,f1score. that's why its concluded that naive bayes and logistic regression are the better performance model for the news group dataset.

talking about the algorithm performance for text classification is totally based on the different dataset. here focusing on the result of the classifiers logistic regression is working best with the 20newsgroups dataset.and evaluating the model performance score is also best for doing the text classification. the best performer overall with the highest accuracy, precision, recall and f1 score values. however the performance of the naive bayes and decison tree is quite low compared to logisitic  regression.

the main part is that performance of the classifier is totally based on the specific dataset and problem is beign solved, so its always the good idea to perform multiple classifier to check which model is working perfectly or better in performance.


All in all, logistic regression model is best for text classification using overall, accuracy score, precision, recall and f1 score. In contrast decision tree and naive bayes is performing low compared to logisitic regression algorithm.